pilot/pkg/simulation/traffic.go

```diff
@@ -0,0 +1,438 @@
+		default:
+			sim.t.Fatalf("unknown route path type")
+		}
+
+		// TODO this only handles path - we need to add headers, query params, etc to be complete.
+
+		return r
+	}
+	return nil
+}
+
+func (sim *Simulation) matchVirtualHost(rc *route.RouteConfiguration, host string) *route.VirtualHost {
+	// Exact match
+	for _, vh := range rc.VirtualHosts {
+		for _, d := range vh.Domains {
+			if d == host {
+				return vh
+			}
+		}
+	}
+	// prefix match
+	var bestMatch *route.VirtualHost
+	longest := 0
+	for _, vh := range rc.VirtualHosts {
+		for _, d := range vh.Domains {
+			if d[0] != '*' {
+				continue
+			}
+			if len(host) >= len(d) && strings.HasSuffix(host, d[1:]) && len(d) > longest {
+				bestMatch = vh
+				longest = len(d)
+			}
+		}
+	}
+	if bestMatch != nil {
+		return bestMatch
+	}
+	// Suffix match
+	longest = 0
+	for _, vh := range rc.VirtualHosts {
+		for _, d := range vh.Domains {
+			if d[len(d)-1] != '*' {
+				continue
+			}
+			if len(host) >= len(d) && strings.HasPrefix(host, d[:len(d)-1]) && len(d) > longest {
+				bestMatch = vh
+				longest = len(d)
+			}
+		}
+	}
+	if bestMatch != nil {
+		return bestMatch
+	}
+	// wildcard match
+	for _, vh := range rc.VirtualHosts {
+		for _, d := range vh.Domains {
+			if d == "*" {
+				return vh
+			}
+		}
+	}
+	return nil
+}
+
+// Follow the 8 step Sieve as in
+// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener_components.proto.html#config-listener-v3-filterchainmatch
+// The implementation may initially be confusing because of a property of the
+// Envoy algorithm - at each level we will filter out all FilterChains that do
+// not match. This means an empty match (`{}`) may not match if another chain
+// matches one criteria but not another.
+func (sim *Simulation) matchFilterChain(chains []*listener.FilterChain, input Call) (*listener.FilterChain, error) {
+	chains = filter(chains, func(fc *listener.FilterChainMatch) bool {
+		return fc.GetDestinationPort() == nil
+	}, func(fc *listener.FilterChainMatch) bool {
+		return int(fc.GetDestinationPort().GetValue()) == input.Port
+	})
+	chains = filter(chains, func(fc *listener.FilterChainMatch) bool {
+		return fc.GetPrefixRanges() == nil
+	}, func(fc *listener.FilterChainMatch) bool {
+		ranger := cidranger.NewPCTrieRanger()
+		for _, a := range fc.GetPrefixRanges() {
+			_, cidr, err := net.ParseCIDR(fmt.Sprintf("%s/%d", a.AddressPrefix, a.GetPrefixLen().GetValue()))
+			if err != nil {
+				sim.t.Fatal(err)
+			}
+			if err := ranger.Insert(cidranger.NewBasicRangerEntry(*cidr)); err != nil {
+				sim.t.Fatal(err)
+			}
+		}
+		f, err := ranger.Contains(net.ParseIP(input.Address))
+		if err != nil {
+			sim.t.Fatal(err)
+		}
+		return f
+	})
+	chains = filter(chains, func(fc *listener.FilterChainMatch) bool {
+		return fc.GetServerNames() == nil
+	}, func(fc *listener.FilterChainMatch) bool {
+		sni := host.Name(input.Sni)
+		for _, s := range fc.GetServerNames() {
+			if sni.SubsetOf(host.Name(s)) {
```

Comment:
nit.. add a todo to allow wildcard based sni match

Reply:
subsetOf is doing wildcard match